server_test.go

```diff
@@ -1,12 +1,77 @@
 package plugin
 
 import (
+	"context"
 	"io/ioutil"
 	"net"
 	"os"
 	"testing"
+	"time"
 )
 
+func TestServer_testMode(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	ch := make(chan *ReattachConfig, 1)
+	closeCh := make(chan struct{})
+	go Serve(&ServeConfig{
+		HandshakeConfig: testHandshake,
+		Plugins:         testGRPCPluginMap,
+		GRPCServer:      DefaultGRPCServer,
+		Test: &ServeTestConfig{
+			Context:          ctx,
+			ReattachConfigCh: ch,
+			CloseCh:          closeCh,
+		},
+	})
+
+	// We should get a config
+	var config *ReattachConfig
+	select {
+	case config = <-ch:
+	case <-time.After(2000 * time.Millisecond):
+		t.Fatal("should've received reattach")
+	}
+	if config == nil {
+		t.Fatal("config should not be nil")
+	}
+
+	// Connect!
+	c := NewClient(&ClientConfig{
+		Cmd:              nil,
+		HandshakeConfig:  testHandshake,
+		Plugins:          testGRPCPluginMap,
+		Reattach:         config,
+		AllowedProtocols: []Protocol{ProtocolGRPC},
+	})
+	client, err := c.Client()
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+
+	// Pinging should work
+	if err := client.Ping(); err != nil {
+		t.Fatalf("should not err: %s", err)
+	}
+
+	// Kill which should do nothing
+	c.Kill()
+	if err := client.Ping(); err != nil {
+		t.Fatalf("should not err: %s", err)
+	}
+
+	// Canceling should cause an exit
+	cancel()
+	<-closeCh
+	if err := client.Ping(); err == nil {
+		t.Fatal("should error")
+	}
+
+	// Try logging, this should show out in tests. We have to manually verify.
```

Comment:
I think at the start of the test we could persist `os.Stdout.Name()` and then compare it here. That wouldn't necessarily catch the race condition I noted, but it would catch other errors without relying on manual verification, I think.

Reply:
We can't use `Name` at the moment because we overwrite it with a multiwriter, unless I'm missing something. :(

Reply:
In my test the multiwriter `Name()` returns `|1`, but I'm not sure this is important.